util/dvsim/SimCfg.py

```diff
@@ -313,64 +313,52 @@ def _print_list(self):
 
     def _create_build_and_run_list(self):
-        # Walk through the list of items to run and create the build and run
-        # objects.
-        # Allow multiple regressions to run as long as the do not enable
-        # sim_modes or run_modes
-        def get_overlapping_tests(tests, run_list_names):
-            overlapping_tests = []
-            for test in tests:
-                if test.name in run_list_names:
-                    overlapping_tests.append(test)
-            return overlapping_tests
-
-        def prune_items(items, marked_items):
-            pruned_items = []
-            for item in items:
-                if item not in marked_items:
-                    pruned_items.append(item)
-            return pruned_items
-
-        # Check if there are items to run
-        if self.items == []:
-            log.error(
-                "No items provided for running this simulation / regression")
-            sys.exit(1)
+        '''Generates a list of deployable objects from the provided items.
 
-        items_list = self.items
-        run_list_names = []
-        marked_items = []
-        # Process regressions first
-        for regression in self.regressions:
-            if regression.name in items_list:
-                overlapping_tests = get_overlapping_tests(
-                    regression.tests, run_list_names)
-                if overlapping_tests != []:
-                    log.error(
-                        "Regression \"%s\" added for run contains tests that overlap with "
-                        "other regressions added. This can result in conflicting "
-                        "build / run_opts to be set causing unexpected results.",
-                        regression.name)
-                    sys.exit(1)
-
-                self.run_list.extend(regression.tests)
-                # Merge regression's build and run opts with its tests and their
-                # build_modes
-                regression.merge_regression_opts()
-                run_list_names.extend(regression.test_names)
-                marked_items.append(regression.name)
-        items_list = prune_items(items_list, marked_items)
-
-        # Process individual tests
-        for test in self.tests:
-            if test.name in items_list:
-                overlapping_tests = get_overlapping_tests([test],
-                                                          run_list_names)
-                if overlapping_tests == []:
-                    self.run_list.append(test)
-                    run_list_names.append(test.name)
-                    marked_items.append(test.name)
-        items_list = prune_items(items_list, marked_items)
+        Tests to be run are provided with --items switch. These can be glob-
+        style patterns. This method finds regressions and tests that match
+        these patterns.
+        '''
+        def _match_items(items: list, patterns_matched: dict, patterns: list):
```

Comment:
I don't think we actually need both `patterns_matched` and `patterns` here, since `patterns` is just the keys of `patterns_matched`. Maybe a bit easier to drop the latter?

Alternatively, and I think this is a bit cleaner, drop the `patterns_matched` argument, change `_match_items` to return a tuple `(hits, matched)` where `hits` is the list of matching items (as you have already) and `matched` is a `set()` of strings for the patterns that matched. Then we can take unions of those sets and finally do a set difference to check whether there were any patterns that didn't match anything.

The reason that it's a bit cleaner is that `_match_items` no longer mutates any of its arguments, which makes the code easier to read.

Reply:
SG, done.